### ISSUE TITLE
About Calculating the Affinity

### ISSUE THREAD

**Issue by dongho-Han** [hkchengrex/STCN#156]
Hello.
I'm trying to understand the logic of your code, and got curious about calculating the affinity.
I think affinity is calculated in training / evaluation session, in different code line.
**In Evaluation**: It's calculated in the Memorybank class' **_global_matching function**(https://github.com/hkchengrex/STCN/blob/eb83102e1957d9f21fb268ab202dad72ef06a1b0/inference_memory_bank.py#L29).
**In Training**: It's calculated in the Memoryreader class' **get_affinity function**(https://github.com/hkchengrex/STCN/blob/eb83102e1957d9f21fb268ab202dad72ef06a1b0/model/network.py#L43).

Based on your paper(thanks to the supplementary), I **understood the evaluation code's description** thoroughly, using top_k(https://github.com/hkchengrex/STCN/blob/eb83102e1957d9f21fb268ab202dad72ef06a1b0/inference_memory_bank.py#L38).

However, I **can't easily understand the training stage's code in a logical way**, especially 
        **x_exp = torch.exp(affinity - maxes)**
this line(https://github.com/hkchengrex/STCN/blob/eb83102e1957d9f21fb268ab202dad72ef06a1b0/model/network.py#L56).
I think until L52, it's almost same with evaluation code. However, after that, it doesn't use top_k methodology, but **get the minus with maxes without explanation**.

Could you share your ideas for this? Any explanation will be helpful!

**Comment by hkchengrex** [comment]
https://stackoverflow.com/questions/42599498/numerically-stable-softmax

**Comment by dongho-Han** [comment]
Dooray! 메일 발송 실패 안내



  
  
    
    
  
  
    
메일 발송
실패 안내
    
  
  
    
        
          
***@***.***) 
          님께
보낸
메일이
전송되지
못하였습니다.
        
        
          실패 사유를 확인해보세요.
        
    
  
  
    
      
        
          
            * 받는 사람 : 
***@***.***) 
          
          
            * 발송 시간 : 
2024-03-29T01:23:30
          
          
            * 메일 제목 : 
Re: [hkchengrex/STCN] About Calculating the Affinity (Issue #156)
          
          
            
              
                * 실패 사유 : 
              
              
                
대외비/기밀 메일은 외부 계정으로 발송이 제한되어 일부 메일 발송 실패(수신자가 전달 설정한 경우 포함)
                
              
            
                      
          
        
      
    
  
  
    
        
          이 메일은 발신전용으로 회신되지 않습니다.
          더 궁금하신 사항은
          ***@***.***
          으로 문의해 주시기 바랍니다.
        
    
  
  
    
        © Dooray!.

**Comment by dongho-Han** [comment]
Dooray! 메일 발송 실패 안내



  
  
    
    
  
  
    
메일 발송
실패 안내
    
  
  
    
        
          
***@***.***) 
          님께
보낸
메일이
전송되지
못하였습니다.
        
        
          실패 사유를 확인해보세요.
        
    
  
  
    
      
        
          
            * 받는 사람 : 
***@***.***) 
          
          
            * 발송 시간 : 
2024-03-29T01:23:29
          
          
            * 메일 제목 : 
Re: [hkchengrex/STCN] About Calculating the Affinity (Issue #156)
          
          
            
              
                * 실패 사유 : 
              
              
                
대외비/기밀 메일은 외부 계정으로 발송이 제한되어 일부 메일 발송 실패(수신자가 전달 설정한 경우 포함)
                
              
            
                      
          
        
      
    
  
  
    
        
          이 메일은 발신전용으로 회신되지 않습니다.
          더 궁금하신 사항은
          ***@***.***
          으로 문의해 주시기 바랍니다.
        
    
  
  
    
        © Dooray!.

**Comment by dongho-Han** [comment]
I thought there would be some reasons, but now I understand that it is for numerical issues.
Thanks for sharing :)
